### ISSUE TITLE
Generic wrapper can't be parametrised with a concrete type meaningfully

### ISSUE THREAD

**Issue by cognivore** [golang/go#48108]
# Technical summary

<pre>
λ go version
go version devel go1.18-faf9c7d8fe Wed Sep 1 03:50:19 2021 +0000 linux/amd64
</pre>

This issue is with generics and is *not* reproducable with the latest release.

<details><summary><code>go env</code> Output</summary><br><pre>
λ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/sweater/.cache/go-build"
GOENV="/home/sweater/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/sweater/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/sweater/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/home/sweater/github/goroot"
GOSUMDB="sum.golang.org"
GOTMPDIR=""     
GOTOOLDIR="/home/sweater/github/goroot/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="devel go1.18-faf9c7d8fe Wed Sep 1 03:50:19 2021 +0000"
GCCGO="gccgo"
AR="ar" 
CC="gcc"
CXX="g++"       
CGO_ENABLED="1"
GOMOD="/home/sweater/github/writing/0001-type-systems-take-over/go/g/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build1099728473=/tmp/go-build -gno-record-gcc-switches"
</pre></details>

# Minimal example

Trying to make a "newtype" that wraps a value of type `T`:

```
λ cat g.go 
package main

import (
        "fmt"
)

type ExprConst[T any] struct {
        Const T
}

type Evaler[T any] interface {
        Eval() T
}

func (e ExprConst[bool]) Eval() int {
        var x bool = e.Const
        fmt.Printf("x: Type: %T Value: %v\n", x, x)
        var t = true
        if x == t {
                return 1
        } else {
                return 0
        }
}

func main() {
        var x int = (ExprConst[bool]{ true }).Eval()
        if x == 1 {
                fmt.Println("All good")
        } else {
                fmt.Println("Oopsie")
        }
}
```

# Expected behaviour

`go run .` prints `All good`.

# Observed behaviour

```
λ go run .
# doma.dev/g
./g.go:19:5: invalid operation: cannot compare x == t (mismatched types bool and bool)
```

# Side-note

Analogous error happens if I parametrise ExprConst with integer type too.

**Comment by griesemer** [comment]
Here's a minimal reproducer:
```Go
package p

type T[P any] struct {
	x P
}

func (e T[bool]) m() {
	if e.x == true {
	}
}
```
Compiling this produces:
```
x.go:8:5: invalid operation: cannot compare e.x == true (mismatched types bool and untyped bool)
```
Note that this is a _correct_ error message. The mistake is in the source code: The method `m` is a method on the parameterized type `T`. Method receivers must declare the same number of type parameters as the receiver base type `T` so that those type parameters are available in the method. This is is _not_ an instantiation. Specifically, this is not an instantiation with the type `bool`. The type parameter name can be any valid name. Typically it will be the same as in the type `T` (so `P` in this case), but in this example you have chosen `bool`. This effectively redeclares the type `bool` for this method to be a type parameter (with constraint `any`). The type of `e.x` is still just `P` (or `bool` which is not the same as the predeclared type `bool` because you redeclared `bool` locally).

This works, but is perhaps not what you intended:
```Go
package p

type T[P any] struct {
	x P
}

func f(e T[bool]) {
	if e.x == true {
	}
}
```

And finally, your version can be made to work with a suitable constraint:
```Go
package p

type T[P interface{ ~bool }] struct {
	x P
}

func (e T[P]) m() { // <<< the name of the type parameter, here P is declared for this method
	if e.x == true {
	}
}
```

In summary, there is no bug here. Closing as working as intended.

**Comment by cognivore** [comment]
@griesemer oh wow, thank you so much for a quick response and for clarifications. I kick myself for not coming up with the correct interpretation (reassigning `bool` identifier to mean `T`)!

I wonder, though, coming back to my intention of simulating Haskell's `newtype ExprConst a = { unExprConst :: a }` together with `exprConstCoerce :: ExprConst a -> (a -> b) -> b`?

While thinking about it, I've found [this bit in the generics proposal](https://go.googlesource.com/proposal/+/refs/heads/master/design/43651-type-parameters.md#methods-may-not-take-additional-type-arguments). Do I understand correctly that there is a way to write an `ExprConstCoerce` function, then for each `a` used in the program manually add it into constraint for the `ExprConst` struct, and then despatch the type in `Eval()`? Is there a static solution that isn't closed wrt possible types of `ExprConst`?

Finally, what would the community suggest as the most idiomatic and open way to achieve newtype wrappers? Typing them out on a case by case basis, so -- in my example -- `ExprConst` would become `ExprBool` and when I'll need to add a wrapper around `int`, I would write `ExprInt`?..

Sorry for a barrage of questions and thank you again for the response! 🙇

**Comment by griesemer** [comment]
It is correct that methods cannot have their own type parameters. Even though we can type-check that, we don't know to implement it efficiently in the presence of interfaces (which would also have methods with corresponding type parameters).

You could add all the types you care about in the constraint, but I'm not sure I fully understand what you're attempting to do (sadly, I am only superficially familiar with Haskel). I'm also not "the community" so I suggest you ask on one of the usual Go channels.

Finally, not everything needs to be done with type parameters. Sometimes an empty interface with suitable accessors (and type assertions) is simpler and more efficient. It may be "less type safe" in theory, but in practice it can be exactly as safe as you want it to be. Specifically, it's likely a mistake to try to translate Haskel code into equivalent Go code - Haskel has mechanisms that Go doesn't have and vice versa.

Thanks.

**Comment by cognivore** [comment]
Thank you very much for your work, @griesemer.

I'm humbled that you have answered both collections of questions. It has helped me greatly in my research.

I was merely trying to prove or disprove that generic Go as implemented sovles the expression problem.

**Comment by griesemer** [comment]
> I was merely trying to prove or disprove that generic Go as implemented solves the expression problem.

Got it. My understanding is that we're not quite there yet with the current design of Go.
